### ISSUE TITLE
About writing tests with functions returning AST objects

### ISSUE THREAD

**Issue by lazarljubenovic** [dsherret/ts-simple-ast#382]
**Is your feature request related to a problem? Please describe.**

Writing a test for a function which returns the following structure.

```ts
interface RouteDefinition {
  path: string
  component: ClassDeclaration
  children?: RouteDefinition[]
}
```

This is how I'm trying to test it.

```ts
  it(`works when there is no nesting`, () => {
    const content = stripIndent`
      class Foo {}
      class Bar {}
      
      export default [
        {
          path: 'foo',
          component: Foo,
        },
        {
          path: 'bar',
          component: Bar,
        },
      ]
    `
    const file = createFile(content)
    const actual = getDefinitionFromFile(createFile(content))
    const fooClassDeclaration = file.getClassOrThrow('Foo')
    const barClassDeclaration = file.getClassOrThrow('Bar')
    const expected: typeof actual = [
      {
        path: 'foo',
        component: fooClassDeclaration,
      },
      {
        path: 'bar',
        component: barClassDeclaration,
      },
    ]
    expect(actual).toEqual(expected)
  })
```

(The function in question is `getDefinitionFromFile`).

The problem is, the internal structure of things change and messes up Jasmine's equality comparison:

![image](https://user-images.githubusercontent.com/7661457/43591165-f7dc7e82-9672-11e8-96b3-1f0367966637.png)

**Describe the solution you'd like**

Some of the ideas:

- Think about this and cover this in the docs with an "official" idea of how to write these tests.
- Add some helper functions or official plugins for common testing libraries for comparing equality of symbols.

**Describe alternatives you've considered**

For my small use-case it was enough to compare the classes by their name.

```ts
function classDeclarationCustomEquality (first: any, second: any): boolean | void {
  if (first instanceof ClassDeclaration && second instanceof ClassDeclaration) {
    return first.getName() == second.getName()
  }
}

  beforeEach(() => {
    jasmine.addCustomEqualityTester(classDeclarationCustomEquality)
  })
```

**Comment by lazarljubenovic** [comment]
Oooh boy I had

```ts
const file = createFile(content)
const actual = getDefinitionFromFile(createFile(content))
```

instead of

```ts
const file = createFile(content)
const actual = getDefinitionFromFile(file)
```

This is why so many things were different, it was a whole different project. 

I guess this can be closed then... A question, though: the structures are guaranteed to remain the same as long as we only query the AST?

**Comment by dsherret** [comment]
I recommend just comparing the node texts in the objects instead of the nodes themselves—so map those `RouteDefinition` objects to a new object that contains the node texts and not the nodes. When something is wrong, it's easy to tell what. Alternatively, once implemented, probably #45 would be useful for writing tests.

Each time after manipulating, the compiler API nodes that are created will be slightly different because the compiler will generate new `id` property values for the nodes. Also, comparing wrapped node objects directly isn't so great because they contain more information than what needs to be compared for equality.

**Comment by lazarljubenovic** [comment]
Thanks!
